core/Network/TLS/Handshake/Server.hs

```diff
@@ -315,9 +315,24 @@ doHandshake sparams mcred ctx chosenVersion usedCipher usedCompression clientSes
             usingHState ctx $ setDHPrivate priv
             return serverParams
 
+        decideHash sigAlg = do
+            usedVersion <- usingState_ ctx getVersion
+            case usedVersion of
+              TLS12 -> do
+                  let sHashSigs = supportedHashSignatures $ ctxSupported ctx
+                      cHashSigs = case extensionLookup extensionID_SignatureAlgorithms exts >>= extensionDecode False of
+                          Just (SignatureAlgorithms hss) -> hss
+                          Nothing                        -> []
+                      hashSigs = cHashSigs `intersect` sHashSigs
+                  case filter ((==) sigAlg . snd) hashSigs of
+                      []  -> error ("no hash signature for " ++ show sigAlg)
+                      x:_ -> return $ Just (fst x)
```

Comment:
Another potential issue is that at this point we seem to have already chosen a cipher, but in fact that choice should have been constrained by the client's list of HashSig pairs, so that by the time we get here we're sure to find a match in the hash sig alg list.  Have not check whether that is true.

Finally, as the code gets more complex it is increasingly important to add comments that explain what's going on.  What assumptions the function makes and expected behaviour, plus anything else not obvious.  If the code is to remain maintainable, the internal documentation must improve faster than the code.

In some cases, as enough new requirements begin to stress the original design, forcing possibly complicated data flows, it may make sense to refactor, and not just make incremental additions.

I am not saying this particular change makes that necessary, but rather encouraging you to think more strategically in the interest of keeping the implementation clean in the long run...